proposals/0836-lean-core-jsc.md

```diff
@@ -0,0 +1,33 @@
+---
+title: Extract JSC (JavaScriptCore) from React Native Core
+author:
+- Kudo Chien <kudo@expo.dev>
+date: 2024-11-28
+---
+
+# RFC0836: Extract JSC (JavaScriptCore) from React Native Core
+
+## Summary
+
+This RFC proposes a Lean Core effort to extract JSC (JavaScriptCore) from React Native's core.
+
+## Proposed Changes and Timeline
+
+- Version 0.77: Remove the [`jsc-android` npm dependency](https://github.com/facebook/react-native/blob/2d337efc23c662143b3f39a6c994d80fec047054/packages/react-native/package.json#L132) from React Native and instead download the artifact from Maven Central. When using Hermes, downloading the unused npm dependency is unnecessary and removing it saves about 32 MB of uncompressed size.
+- Version 0.78 or 0.79:  Introduce a new `react-native-javascriptcore` npm package with a newer version of JavaScriptCore for Android. In this version, we will only provide the [Intl variant](https://github.com/react-native-community/jsc-android-buildscripts/blob/9c61fece4753902a2cd6d29dfa46b7b521f0c821/README.md#international-variant), as newer JavaScriptCore versions do not support disabling Intl. The existing JSCRuntime from the core should still function with this version but will display a deprecation warning during build time. A challenge will be managing the core JSCRuntime alongside the third-party JSCRuntime from `react-native-javascriptcore`; further study is required.
+- Version 0.80: Completely remove JavaScriptCore from the core:
+  - Eliminate `JSCRuntime` code.
+  - Deprecate Gradle's `hermesEnabled` and Podfile's `hermes_enabled`, making them no-ops.
```

Comment:
This has hidden complexity. Check here:
https://github.com/search?type=code&q=repo%3Afacebook%2Freact-native+%22hermesEnabled%22+language%3AKotlin&l=Kotlin

If we deprecate `hermesEnabled` and set it to `true`, we'll always produce a bundle that is not processed by `hermesc` which jsc won't be able to consume.

So in a sense we need to refactor the build logic on Android & iOS so that you can still decide if `hermesc` should execute or not.

Moreover on Android, we selectively remove `.so` based on the `hermesEnabled` so that logic will have to be adopted as well.